containers/ecr-viewer/src/app/api/services/core_database_repo.ts

```diff
@@ -0,0 +1,337 @@
+
+import {
+  ECRConditions,
+  NewECRConditions,
+  ECRConditionsUpdate,
+  ECRRuleSummaries,
+  NewECRRuleSummaries,
+  ECRRuleSummariesUpdate,
+  ECR,
+  NewECR,
+  ECRUpdate,
+} from "./core_types";
+import { Core } from "./core_types";
+import { db } from "./database";
+
+// ECR_DATA
+
+/**
+ * Finds an eICR by its ID
+ * @async
+ * @function findEcrById
+ * @param id - thd ID of the eCR to be looked up
+ * @returns an eICR object
+ */
+export async function findEcrById(id: string | null): Promise<ECR | undefined> {
+  if (!id) {
+    throw new Error("eICR ID is required.");
+  }
+  try {
+    return await (db as Kysely<Core>)
+      .selectFrom("ecr_data")
+      .where("eICR_ID", "=", id)
+      .selectAll()
+      .executeTakeFirst();
+  } catch (error) {
+    console.error(error);
+  }
+}
+
+/**
+ * Finds an eICR by its criteria
+ * @async
+ * @function findEcr
+ * @param criteria - the Partial<ECR> used to filter against
+ * @returns an eICR object
+ */
+export async function findEcr(criteria: Partial<ECR> | null): Promise<ECR[]> {
+  let query = (db as Kysely<Core>).selectFrom("ecr_data");
+
+  if (!criteria || criteria === null) {
+    throw new Error("eICR Criteria is required.");
+  }
+
+  for (const criterium of Object.keys(criteria) as (keyof ECR)[]) {
+    if (criteria[criterium] !== undefined && criteria[criterium] !== null) {
+      query = query.where(criterium, "=", criteria[criterium]);
+    }
+  }
+
+  return await query.selectAll().execute();
+}
+
+/**
+ * Creates an eICR object
+ * @async
+ * @function createEcr
+ * @param ecr - the NewECR to be persisted
+ * @returns the created eICR object
+ */
+export async function createEcr(ecr: NewECR | null): Promise<ECR | undefined> {
+  if (!ecr || ecr === null) {
+    throw new Error("eICR Data is required.");
+  }
+  try {
+    return await (db as Kysely<Core>)
+      .insertInto("ecr_data")
+      .values(ecr)
+      .returningAll()
+      .executeTakeFirstOrThrow();
+  } catch (error) {
+    console.error(error);
+  }
+}
+
+/**
+ * Updates an eICR object
+ * @async
+ * @function updateEcr
+ * @param eICR_ID - the eICR_ID of the eCR to be updated
+ * @param updateWith - the ECRUpdate to be applied to the existing record
+ * @returns the updated eICR object
+ */
+export async function updateEcr(
+  eICR_ID: string | null,
+  updateWith: ECRUpdate,
+): Promise<void> {
+  await (db as Kysely<Core>)
+    .updateTable("ecr_data")
+    .set(updateWith)
+    .where("eICR_ID", "=", eICR_ID)
+    .execute();
+}
+
+/**
+ * Deletes an eICR object
+ * @async
+ * @function deleteEcr
+ * @param eICR_ID - the eICR_ID of the record to be deleted
+ * @returns the deleted eICR object
+ */
+export async function deleteEcr(
+  eICR_ID: string | null,
+): Promise<ECR | undefined> {
+  const ecr = await findEcrById(eICR_ID);
+
+  if (!ecr) {
+    return undefined;
+  }
+
+  await (db as Kysely<Core>)
+    .deleteFrom("ecr_data")
+    .where("eICR_ID", "=", eICR_ID)
+    .execute();
```

Comment:
why do we need to find it before deleting it? is there a latent race condition here where it could be found then deleted before this delete since it's done as two different transactions?